### DISCUSSION TITLE
how to use UserNotificationsDependency delegate

### DISCUSSION THREAD

**Discussion by corysullivan** [tgrapperon/swift-dependencies-additions#16] [Q&A]
Thanks for the great work! This is an excellent addition to the Dependency package and will help clean up my app😀. I did have a question on `UserNotificationsDependency`, specifically how to use the delegate. I have tried using the iso-words implementation [found here](https://github.com/pointfreeco/isowords/blob/bac580f376d68e7b280cf51048e1a95f81873bd0/Sources/ComposableUserNotifications/LiveKey.swift#L8), this version did not set the delegate on the main thread, and consequently was not set in time for cold launches triggered by a push notification. Therefore, I  abandoned this idea. However, it looks like you take a different approach, and I am interested in trying your implementation. Do you have an example illustrating how we should use the delegate in a TCA application? I'm unsure how the delegate events get propagated into a reducer action.

**Comment by mbrandonw** [comment]
Hey @corysullivan, sorry to poke into this thread, but I was curious about your comments about the delegate not being set on the main thread, and so possibly losing the cold launch triggers. I looked into it, and it is true that we call `delegate` inside a task group task:

```swift
        await withThrowingTaskGroup(of: Void.self) { group in
          group.addTask {
            for await event in self.userNotifications.delegate() {
              await send(.userNotifications(event))
            }
          }
```
https://github.com/pointfreeco/isowords/blob/bac580f376d68e7b280cf51048e1a95f81873bd0/Sources/AppFeature/AppDelegate.swift#L32-L37

…however, that endpoint could have been called _outside_ of the effect and then iterated over _inside_ the effect:

```swift
      let notificationEvents = self.userNotifications.delegate()
      return .run { send in
        await withThrowingTaskGroup(of: Void.self) { group in
          group.addTask {
            for await event in notificationEvents {
              await send(.userNotifications(event))
            }
          }
```

I think then things would have worked correctly. I just wanted to mention that because it is subtle. It is totally fine to construct the `AsyncStream` outside the effect. You just have to be inside the effect to iterate over it.

**Reply by corysullivan** [reply]
You're right, this does fix the issue. As a side note I also had to add a reference to the delegate inside the onTermination callback otherwise the delegate was released immediately. But yeah it's working! Thanks
```swift
AsyncStream { continuation in
  let delegate = Delegate(continuation: continuation)
  UNUserNotificationCenter.current().delegate = delegate
  continuation.onTermination = { [delegate] _ in
    // We need to reference the delegate inside the closure otherwise the delegate is released.
    _ = delegate
  }
}
```

**Comment by tgrapperon** [comment]
Hey @corysullivan! I principally implemented this dependency with Vanilla applications in mind for the time being, so I can't provide a direct example of use in TCA right now. It is maybe possible to make the delegate of this dependency something itself configurable (I mean each endpoint, not just the instance like right now). But delegates are quite common, so I'll think a bit more about a way to generalize the solution.
Having some example of use in TCA is a very good idea. I'll add some case studies very soon.

**Reply by nickkohrn** [reply]
Hey @tgrapperon! Have you given additional thought to this? I'm trying to work out how to handle the delegate for `UserNotificationsDependency` in my TCA project, but I'm coming up short of a solution.
